### PR TITLE
Refactor PACT commands

### LINKED ISSUE #159: Refactor PACT commands
## Objective
The main objective is to allow adding a list of signers to the commands, from a KeyPair, also to correct some bugs, and allow sending an EnvData type to the commands.
## Checklist

- [x] Allow a command to be created with only the minimum necessary data
- [x] Set default scheme ED25519 when the key is 32 bytes.
- [x] Create signersList from a KeyPair.
- [x] Allow to create valid commands without keypair

### PR DESCRIPTION
**Pull request by Odraxs**
## Description

Fixes #159 

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality not to work as expected)
- [x] This change requires a documentation update

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented on my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: lib/types/command_payload.ex [hunk 1]
```diff
@@ -89,6 +90,7 @@ defmodule Kadena.Types.CommandPayload do
 
   @spec validate_nonce(nonce :: nonce()) :: validation()
   defp validate_nonce(nonce) when is_binary(nonce), do: {:ok, nonce}
+  defp validate_nonce(nonce) when is_nil(nonce), do: {:ok, ""}
```
Comment (juanhurtado10) [inline lib/types/command_payload.ex]: @Odraxs My suggestion: transform the nonce into a string `to_string/1`

File: lib/types/command_payload.ex [hunk 2]
```diff
@@ -79,6 +79,7 @@ defmodule Kadena.Types.CommandPayload do
 
   @spec validate_meta(meta :: meta()) :: validation()
   defp validate_meta(%MetaData{} = meta), do: {:ok, meta}
+  defp validate_meta(meta) when is_nil(meta), do: {:ok, %MetaData{}}
```
Comment (juanhurtado10) [inline lib/types/command_payload.ex]: @Odraxs instead of using a guard you can pattern match directly

File: lib/cryptography/sign/default.ex [hunk 3]
```diff
@@ -19,6 +19,13 @@ defmodule Kadena.Cryptography.Sign.Default do
   @edwards_curve_ed :ed25519
 
   @impl true
+  def sign(msg, nil) do
+    with bin_msg <- blake2b_hash(msg, byte_size: 32),
+         hash <- url_encode64(bin_msg) do
+      {:ok, SignCommand.new(hash: hash)}
```
Comment (juanhurtado10) [inline lib/cryptography/sign/default.ex]: @Odraxs this function is not really clear to me, no matter what it will return an error tuple.

image

Are you really signing a message? (The answer to this question will define the responsibility of your function)

File: lib/pact/command/cont_command.ex [hunk 4]
```diff
@@ -215,7 +218,8 @@ defmodule Kadena.Pact.ContCommand do
           keypairs: keypairs
         } = cmd_request
       ) do
-    with {:ok, payload} <- create_payload(cmd_request),
+    with {:ok, cmd_request} <- set_signers_from_keypair(cmd_request),
+         {:ok, payload} <- create_payload(cmd_request),
```
Comment (juanhurtado10) [inline lib/pact/command/cont_command.ex]: @Odraxs Let's think about this use case: a Command that has added signers and key pairs ..... how would the code handle this?

File: lib/pact/command/cont_command.ex [hunk 5]
```diff
@@ -227,6 +231,36 @@ defmodule Kadena.Pact.ContCommand do
...
+
+  @spec set_signers(t()) :: t()
+  defp set_signers(%__MODULE__{keypairs: key_pairs} = cmd_request) do
+    Enum.map(key_pairs, fn %KeyPair{pub_key: pub_key, clist: clist} ->
+      Signer.new(pub_key: pub_key, clist: clist)
+    end)
+    |> SignersList.new()
+    |> (&add_signers(cmd_request, &1)).()
+  end
+
```
Comment (juanhurtado10) [inline lib/pact/command/cont_command.ex]: @Odraxs after answering the previous question. It is necessary to update/refactor these functions.

File: lib/pact/command/cont_command.ex [hunk 6]
```diff
@@ -273,6 +307,11 @@ defmodule Kadena.Pact.ContCommand do
 
   @spec sign_commands(signs :: list(), cmd :: json_string_payload(), keypairs()) ::
           valid_sign_commands()
+  defp sign_commands([], cmd, []) do
+    {:ok, sign_command} = Sign.sign(cmd, nil)
+    {:ok, [sign_command]}
+  end
```
Comment (juanhurtado10) [inline lib/pact/command/cont_command.ex]: @Odraxs this code snippet is a bit confusing to me, IMO it can be handled in a better way

**Review by juanhurtado10 — CHANGES_REQUESTED** [review]

File: lib/cryptography/sign/default.ex (hunk 3 above)
Comment (EdwinGuayacan) [reply lib/cryptography/sign/default.ex]: we modify the `SignatureWithHash` type, which is called in the `new/1` function of the `SignCommand`, so that it receives a sign nil and we can obtain the hash.
image

File: lib/pact/command/cont_command.ex (hunk 4 above)
Comment (EdwinGuayacan) [reply lib/pact/command/cont_command.ex]: In that use case, we should add signatures to the existing list, making use of the Key Pairs??

File: lib/cryptography/sign/default.ex (hunk 3 above)
Comment (juanhurtado10) [reply lib/cryptography/sign/default.ex]: @EdwinGuayacan ok, I understood you but IMO, it is not a signature, it is just a hash and a command without a signature. 

Why process it through the `sign/2` function and call it a signature?

File: lib/pact/command/cont_command.ex (hunk 4 above)
Comment (juanhurtado10) [reply lib/pact/command/cont_command.ex]: @EdwinGuayacan Here is my suggestion.

1. Keep the functions for adding keypairs and signers as they are.
2. Every time `add_keypair/2` is invoked, build the signer. (prevent duplicates).
3. That's it.

This will allow you to have both key-pair signers and individual signers

Comment (Odraxs) [reply lib/pact/command/cont_command.ex]: @juanhurtado10 is it possible to add the same keypair several times, is that ok or is it also necessary to avoid the keypair being duplicated when using `add_keypair/2` or do we just need to prevent duplicate signers?

Comment (juanhurtado10) [reply lib/pact/command/cont_command.ex]: @Odraxs 🤔 that's a good question. Let's simplify, for this PR let's allow duplicates for both. Then we create a specific issue to handle duplicates.

File: lib/pact/command/exec_command.ex [hunk 7]
```diff
@@ -180,6 +184,36 @@ defmodule Kadena.Pact.ExecCommand do
...
+       ) do
+    case signers == [] && keypairs != [] do
+      true -> {:ok, set_signers(cmd_request)}
+      false -> {:ok, cmd_request}
+    end
+  end
+
+  @spec set_signers(t()) :: t()
+  defp set_signers(%__MODULE__{keypairs: key_pairs} = cmd_request) do
+    Enum.map(key_pairs, fn %KeyPair{pub_key: pub_key, clist: clist} ->
```
Comment (L-Zuluaga) [inline lib/pact/command/exec_command.ex]:
```suggestion
    key_pairs
   |> Enum.map(fn %KeyPair{pub_key: pub_key, clist: clist} ->
```

File: lib/pact/command/exec_command.ex [hunk 8]
```diff
@@ -220,6 +236,14 @@ defmodule Kadena.Pact.ExecCommand do
 
   @spec sign_commands(signs :: list(), cmd :: json_string_payload(), keypairs()) ::
           valid_sign_commands()
+  defp sign_commands(signs, cmd, []) when signs == [] do
```
Comment (juanhurtado10) [inline lib/pact/command/exec_command.ex]: @EdwinGuayacan @Odraxs you can remove the guard and use pattern matching instead

File: lib/types/command_payload.ex [hunk 9]
```diff
@@ -79,6 +79,7 @@ defmodule Kadena.Types.CommandPayload do
 
   @spec validate_meta(meta :: meta()) :: validation()
   defp validate_meta(%MetaData{} = meta), do: {:ok, meta}
+  defp validate_meta(nil), do: {:ok, %MetaData{}}
```
Comment (juanhurtado10) [inline lib/types/command_payload.ex]: @Odraxs @EdwinGuayacan a better approach would be to initialize the empty metadata structure like this `Metadata.new()`.

File: lib/types/command_payload.ex [hunk 10]
```diff
@@ -89,6 +90,7 @@ defmodule Kadena.Types.CommandPayload do
 
   @spec validate_nonce(nonce :: nonce()) :: validation()
   defp validate_nonce(nonce) when is_binary(nonce), do: {:ok, nonce}
+  defp validate_nonce(nil), do: {:ok, to_string(nil)}
   defp validate_nonce(_nonce), do: {:error, [nonce: :not_a_string]}
```
Comment (juanhurtado10) [inline lib/types/command_payload.ex]: @Odraxs @EdwinGuayacan just replace these three function clauses by

```
defp set_nonce(nonce), do: {:ok, to_string(noce)}
```

File: lib/types/meta_data.ex [hunk 11]
```diff
@@ -26,7 +26,12 @@ defmodule Kadena.Types.MetaData do
           chain_id: chain_id()
         }
 
-  defstruct [:creation_time, :ttl, :gas_limit, :gas_price, :sender, :chain_id]
+  defstruct creation_time: 0,
+            ttl: 0,
+            gas_limit: 0,
+            gas_price: 0,
+            sender: "",
+            chain_id: %ChainID{id: ""}
```
Comment (juanhurtado10) [inline lib/types/meta_data.ex]: @Odraxs @EdwinGuayacan Why is this required?

**Review by juanhurtado10 — CHANGES_REQUESTED** [review]

File: lib/types/meta_data.ex (hunk 11 above)
Comment (EdwinGuayacan) [reply lib/types/meta_data.ex]: In the tests we did to the local endpoint, we saw that it did not allow null values in the MetaData attributes:
`{"cmd":"{\"meta\":{\"chainId\":null,\"creationTime\":null,\"gasLimit\":null,\"gasPrice\":null,\"sender\":null,\"ttl\":null},\"networkId\":null,\"nonce\":\"\",\"payload\":{\"exec\":{\"code\":\"(+ 5 6)\",\"data\":null}},\"signers\":[]}","hash":"EOJAI1UaQZ3T4Zzv9V95Rhu89dI7gQqltTVA_qdHr4w","sigs":[]}`

image

Comment (juanhurtado10) [reply lib/types/meta_data.ex]: @EdwinGuayacan ok ok I got you ... let's use 0 as a default value for the ChanID.

File: lib/types/meta_data.ex [hunk 12]
```diff
@@ -26,7 +26,15 @@ defmodule Kadena.Types.MetaData do
           chain_id: chain_id()
         }
 
-  defstruct [:creation_time, :ttl, :gas_limit, :gas_price, :sender, :chain_id]
+  defstruct creation_time: 0,
+            ttl: 0,
+            gas_limit: 0,
+            gas_price: 0,
+            sender: "",
+            chain_id: %ChainID{id: "0"}
```
Comment (juanhurtado10) [inline lib/types/meta_data.ex]: @Odraxs @EdwinGuayacan please use `ChainID.new()`

File: lib/types/meta_data.ex [hunk 13]
```diff
@@ -26,7 +26,15 @@ defmodule Kadena.Types.MetaData do
...
+
+  @impl true
+  def new, do: %__MODULE__{}
 
   @impl true
   def new(args) when is_list(args) do
```
Comment (juanhurtado10) [inline lib/types/meta_data.ex]: @Odraxs @EdwinGuayacan This is not really necessary, you can use a default value for the arguments

```elixir
def new (args \\ [])
def new(args) when is_list(args) do
  ...
end
```

File: lib/types/spec.ex [hunk 14]
```diff
@@ -6,5 +6,8 @@ defmodule Kadena.Types.Spec do
   @type reasons :: Keyword.t()
   @type error :: {:error, reasons()}
 
+  @callback new() :: struct()
   @callback new(any()) :: struct() | error()
+
+  @optional_callbacks new: 0
```
Comment (juanhurtado10) [inline lib/types/spec.ex]: @EdwinGuayacan @Odraxs Following the approach mentioned above, this callback is not necessary.

**Review by juanhurtado10 — CHANGES_REQUESTED** [review]

File: lib/pact/command/cont_command.ex [hunk 15]
```diff
@@ -74,7 +74,7 @@ defmodule Kadena.Pact.ContCommand do
     :step,
     :proof,
     :rollback,
-    :signers,
+    signers: [],
```
Comment (miguelnietoa) [inline lib/pact/command/cont_command.ex]: @Odraxs @EdwinGuayacan  `@type signers` indicates that this attribute must be a `SignerList.t()`, but we're setting by default an empty list, do you think this is right? 🤔 Same for ExecCommand...

**Review by miguelnietoa — CHANGES_REQUESTED** [review]

File: lib/pact/command/cont_command.ex [hunk 16]
```diff
@@ -91,7 +91,7 @@ defmodule Kadena.Pact.ContCommand do
     proof = Keyword.get(opts, :proof)
     rollback = Keyword.get(opts, :rollback, %Rollback{})
     keypairs = Keyword.get(opts, :keypairs, [])
-    signers = Keyword.get(opts, :signers, %SignersList{})
+    signers = Keyword.get(opts, :signers, [])
```
Comment (miguelnietoa) [inline lib/pact/command/cont_command.ex]: @Odraxs @EdwinGuayacan  the default value in `new/2` for rollback is a struct, but `set_rollback/2` function does not match this but a boolean... so, this field will be a boolean or a struct?

**Review by miguelnietoa — CHANGES_REQUESTED** [review]

File: lib/pact/command/cont_command.ex (hunk 16 above)
Comment (Odraxs) [reply lib/pact/command/cont_command.ex]: @miguelnietoa Because of this, I discovered that the values `step`, `pact_transaction_hash`, and `rollback` are defined as types even though they are actually just numeric, string, and boolean values, respectively.

I suppose it would be better if we change those methods to use their types?

Comment (miguelnietoa) [reply lib/pact/command/cont_command.ex]: @Odraxs Mmmm given this situation, I think we can use the type structs internally, but the commands could receive the primitive value in _number, string, boolean_, as you mentioned...

I think it can follow the same logic behind `set_network/2` where you provide `:testnet04` but is saved internally as `%NetworkID{id: "testnet04"}`

Is it clear? What do you think?

Comment (EdwinGuayacan) [reply lib/pact/command/cont_command.ex]: Yes, that's the idea, in fact the primitive data is received and the types are created in `Cont Payload` type

File: lib/pact/command/cont_command.ex [hunk 17]
```diff
@@ -38,14 +34,14 @@ defmodule Kadena.Pact.ContCommand do
...
-  @type rollback :: Rollback.t()
+  @type pact_tx_hash :: String.t()
+  @type proof :: String.t() | nil
+  @type rollback :: boolean()
   @type signatures :: SignaturesList.t()
   @type signers :: SignersList.t()
   @type sign_command :: SignCommand.t()
   @type sign_commands :: list(sign_command())
-  @type step :: Step.t()
+  @type step :: number()
```
Comment (miguelnietoa) [inline lib/pact/command/cont_command.ex]: @Odraxs @EdwinGuayacan IMO integer suits better

File: lib/pact/command/cont_command.ex [hunk 18]
```diff
@@ -162,8 +165,11 @@ defmodule Kadena.Pact.ContCommand do
   def add_keypairs({:error, reason}, _keypairs), do: {:error, reason}
 
   @impl true
-  def add_signer(%__MODULE__{signers: nil} = cmd_request, %Signer{} = signer),
-    do: %{cmd_request | signers: SignersList.new([signer])}
+  def add_signer(
+        %__MODULE__{signers: %SignersList{signers: []}} = cmd_request,
+        %Signer{} = signer
+      ),
+      do: %{cmd_request | signers: SignersList.new([signer])}
```
Comment (miguelnietoa) [inline lib/pact/command/cont_command.ex]: @Odraxs @EdwinGuayacan is this function clause matching now required?

File: lib/pact/command/cont_command.ex [hunk 19]
```diff
@@ -174,8 +180,16 @@ defmodule Kadena.Pact.ContCommand do
   def add_signer({:error, reason}, _signer), do: {:error, reason}
 
   @impl true
-  def add_signers(%__MODULE__{} = cmd_request, %SignersList{} = list),
-    do: %{cmd_request | signers: list}
+  def add_signers(
+        %__MODULE__{signers: %SignersList{signers: []}} = cmd_request,
+        %SignersList{} = list
+      ),
+      do: %{cmd_request | signers: list}
```
Comment (miguelnietoa) [inline lib/pact/command/cont_command.ex]: @Odraxs @EdwinGuayacan is this function clause matching now required?

**Review by miguelnietoa — CHANGES_REQUESTED** [review]

**Review by miguelnietoa — APPROVED** [review]

**Review by juanhurtado10 — APPROVED** [review]
